### DISCUSSION TITLE
Netbox in an Jupyter Notebook?

### DISCUSSION THREAD

**Discussion by ziggekatten** [netbox-community/netbox#15517] [Q&A]
We are doing some stuff where it would be of great use to have access to Netbox models (in effect nbshell) and stuff in Jupyter notebooks. 

Anyone that have done this?

**Comment by bitcollector1** [comment]
You got me curious, so I setup the virtual environment and tweaked the notebook as follows   
```
import os
os.environ["DJANGO_ALLOW_ASYNC_UNSAFE"] = "true" 
    
%run -i '/opt/netbox/netbox/manage.py' 'nbshell'
```

<img width="916" alt="Screenshot 2024-03-28 at 22 18 26" src="https://github.com/netbox-community/netbox/assets/50723251/dbc53c40-4da7-4479-abed-3a9298c5f93f">

<img width="638" alt="Screenshot 2024-03-28 at 22 21 05" src="https://github.com/netbox-community/netbox/assets/50723251/396bb6be-0c18-44a1-b59b-b68ffd61f98c">

**Reply by ziggekatten** [reply]
Hmm...clearly I'm not an master of jupyter, and I can't get it to work properly. Get nested nastiness here:
![image](https://github.com/netbox-community/netbox/assets/55629049/89a9bf49-e895-4322-b39e-fa042a0584ee)

Also,

**Reply by bitcollector1** [reply]
There are quite a few articles on the web, I think the hardest part is setting up Jupyter to use the proper virtual environment. 

Did some more messing around today.


 
<img width="893" alt="Screenshot 2024-03-29 at 08 44 40" src="https://github.com/netbox-community/netbox/assets/50723251/ec8587c7-4aed-4a25-ab17-fd03e48bc234">

**Reply by bitcollector1** [reply]
So glad you asked this question and stirred my curiosity. Now I can actually start taking notes for how all this NB Shell stuff works and transition from pynetbox. 

Working directly with the Models gives you a better understanding of how this all works and seems to give you some very interesting ways to get at the data that might not be so obvious from the API wrapper!

**Reply by ziggekatten** [reply]
I thought I made it simple by simply install jupyter in my netbox venv ad start it from there, but it seems that its more to it than that. Spent whole day trying to get this running. no idea how you did it, but I can't simply make it work according to your screenshots. 

Nbshell is such a usefull tool when writing custom scripts. And since we use stuff outside netbox, in my case prototyping texfsm templates and pushing stuff via Pandas, jupyter would really be a nice frontend.

**Reply by bitcollector1** [reply]
you might need to make sure notebook has access to the kernel 

```
ebay@netdevbox:~$ jupyter kernelspec list
Available kernels:
  3.11.0       /usr/share/jupyter/kernels/3.11.0
  3.12         /usr/share/jupyter/kernels/3.12
  3.8.11       /usr/share/jupyter/kernels/3.8.11
  nb_python    /usr/share/jupyter/kernels/nb_python
  python3      /usr/share/jupyter/kernels/python3
```

```
ebay@netdevbox:~$ cat /usr/share/jupyter/kernels/nb_python/kernel.json 
{
 "argv": [
  "/opt/netbox/venv/bin/python3",
  "-m",
  "ipykernel_launcher",
  "-f",
  "{connection_file}"
 ],
 "display_name": "nb_python",
 "language": "python",
 "metadata": {
  "debugger": true
 }
}
```

**Reply by bitcollector1** [reply]
Also, run this from within the activated netbox env
```
python3 -m pip install ipykernel
```

pretty sure all you need are these couple steps!  good luck!

**Reply by ziggekatten** [reply]
Well, i can't see any difference here. Scratching my head a bit. What version of Jupyter are you running? Your logo looks different than mine (7.1.2). 
```
cat /usr/local/share/jupyter/kernels/nb_venv/kernel.json
{
 "argv": [
  "/opt/netbox/venv/bin/python3",
  "-m",
  "ipykernel_launcher",
  "-f",
  "{connection_file}"
 ],
 "display_name": "nb_venv",
 "language": "python",
 "metadata": {
  "debugger": true
 }
}
```

```
jupyter kernelspec list
Available kernels:
  python3    /opt/netbox-3.7.4/venv/share/jupyter/kernels/python3
  nb_venv    /usr/local/share/jupyter/kernels/nb_venv
```

**Reply by bitcollector1** [reply]
It looked like you had it almost working, not sure why it was red? maybe SSL or something? 

I saw the input box for the nbshell in your screenshot. Did you try typing in the following into the in: box 

```
help(IPAddress)
```

<img width="634" alt="Screenshot 2024-03-29 at 15 09 03" src="https://github.com/netbox-community/netbox/assets/50723251/a928a1c8-4d74-40fe-8b3f-e3a775eecbfc">

```

**Reply by bitcollector1** [reply]
you can also run this to for a non-interactive mode
```
%run -i '/opt/netbox/netbox/manage.py' 'nbshell' '-c' 'lsmodels()'
```

finally you also can use the cells normally if you just import the models and start using them


<img width="858" alt="Screenshot 2024-03-29 at 15 12 06" src="https://github.com/netbox-community/netbox/assets/50723251/266ddb19-4c50-4475-8b8f-e6b04fb721cf">

<img width="639" alt="Screenshot 2024-03-29 at 15 12 40" src="https://github.com/netbox-community/netbox/assets/50723251/01bfdf6d-e005-4760-9483-b25207bf0ea6">

**Reply by ziggekatten** [reply]
Yes, I can type in the field and get response. Problem is that it don't use the standard cells, and I cant go back/re execute commands.  It's like stout/in is not redirected to the notebook server
![image](https://github.com/netbox-community/netbox/assets/55629049/7fcc6450-ad7b-4a9c-81a4-8e15dac868b2)

**Reply by bitcollector1** [reply]
yes, you get "stuck" in that mode....it's not advised.  Try just using the models themselves as I showed.  

I had same issues until I figured out how to run the command and quit. And then finally figured out I could use the model objects themselves in the cells once they were loaded :)

**Reply by ziggekatten** [reply]
Yea, it seems that I would need to import the models in an traditional way. The sweet thing with nbshell is that all models are loaded by default. Well, an small function that goes through the netbox.settings is probably the way to go then

**Reply by bitcollector1** [reply]
If you load the proper netbox venv, I believe all the models are available to you.

**Comment by bitcollector1** [comment]
<img width="749" alt="Screenshot 2024-03-29 at 15 15 52" src="https://github.com/netbox-community/netbox/assets/50723251/e8b68373-208d-4f13-9f6a-47f88d933cfb">

**Comment by bitcollector1** [comment]
now I can start documenting all the screen shots from @candlerb that I've been collecting over the years ;)
